File: common/internal_model/src/main/scala/uk/ac/wellcome/models/work/internal/Location.scala
package uk.ac.wellcome.models.work.internal

sealed trait Location {
  val locationType: LocationType
  val accessConditions: List[AccessCondition]
  val license: Option[License]

  def hasRestrictions: Boolean =
    accessConditions.exists(_.hasRestrictions)
}

case class DigitalLocation(
  url: String,
  locationType: DigitalLocationType,
  license: Option[License] = None,
  credit: Option[String] = None,
  linkText: Option[String] = None,
  accessConditions: List[AccessCondition] = Nil
) extends Location

case class PhysicalLocation(
  locationType: PhysicalLocationType,
  label: String,
  license: Option[License] = None,
  shelfmark: Option[String] = None,
  accessConditions: List[AccessCondition] = Nil
) extends Location {

  if (locationType == LocationType.ClosedStores) {
    require(
      label == LocationType.ClosedStores.label,
      s"Label is '$label', but we don't want to expose the layout of our Closed Stores"
    )
  }
}

File: catalogue_graph/src/ingestor/models/shared/location.py
from .access_condition import AccessCondition
from .id_label import Id
from .serialisable import ElasticsearchModel


class Location(ElasticsearchModel):
    location_type: Id
    license: Id | None = None
    access_conditions: list[AccessCondition]


class DigitalLocation(Location):
    url: str
    credit: str | None = None
    link_text: str | None = None


class PhysicalLocation(Location):
    label: str
    shelfmark: str | None = None
